blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc

```diff
@@ -0,0 +1,647 @@
+= Building Audit Logs with Change Data Capture and Stream Processing
+gmorling
+:awestruct-tags: [ discussion, examples, apache-kafka, kafka-streams, featured ]
+:awestruct-layout: blog-post
+
+It is a common requirement for business applications to maintain some form of audit log,
+i.e. a persistent trail of all the changes to the application's data.
+If you squint a bit, a Kafka topic with Debezium data change events is quite similar to that:
+sourced from database transaction logs, it describes all the changes to the records of an application.
+What's missing though is some metadata: why, when and by whom was the data changed?
+In this post we're going to explore how that metadata can be provided and exposed via change data capture (CDC), and how stream processing can be used to enrich the actual data change events with such metadata.
+
+Reasons for maintaining data audit trails are manifold:
+e.g. regulatory requirements may mandate businesses to keep complete historic information of their customer, purchase order, invoice or other data.
+Also for an enterprise's own purposes it can be very useful to be able to analyze why and how certain data has changed, e.g. allowing to improve business processes and analyze errors.
+
+Traditionally, application-side libraries and database triggers have been common means for creating audit trails:
+when a data record is inserted, updated or deleted, they'd maintain specific column(s) in the data tables ("createdBy", "lastUpdated" etc.) and/or copy earlier record versions into some form of history tables.
+This comes with some disadvantages, though:
+
+* metadata columns and triggers add bloat to the model in the database and the corresponding application domain model
+* writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction and thus may cause longer response times of the application
+
+Library-based solutions also oftentimes cannot provide metadata in case of bulk updates and deletes (`UPDATE mytable WHERE ...`),
+as the listeners used to hook them into the persistence framework are not aware of all the affected records.
+Also, by definition, any approach based on an application-side library is prone to miss any data changes done directly in the database, e.g. when running a data load or bypassing the application during an emergency data patch.
```

Comment:
Also here the same holds true for the proposed solution. Since direct changes won't trigger application level meta data insertion which is done in the example code. The only difference is that you can e.g. "block the enrichments" in the stream processor but still there is no such data available under such circumstances.

Reply:
Yes, that's true as far as the metadata information is concerned. The actual change events of course would be captured from the log also for a bulk update/delete. I'll see how I can rephrase that; I was unsure about that myself, but thought I'd get away with it. Seems that not :)

Reply:
I hope it's clearer now with the rewording of this section.